project-auth/src/main/scala/io/renku/projectauth/ProjectAuthService.scala

```diff
@@ -0,0 +1,147 @@
+      connectionConfig: ConnectionConfig,
+      timeout:          Duration = 20.minutes
+  )(implicit renkuUrl: RenkuUrl): Resource[F, ProjectAuthService[F]] =
+    SparqlClient(connectionConfig, timeout).map(c => apply[F](c, renkuUrl))
+
+  def apply[F[_]: MonadThrow](client: SparqlClient[F], renkuUrl: RenkuUrl): ProjectAuthService[F] =
+    new Impl[F](client, renkuUrl)
+
+  private final class Impl[F[_]: MonadThrow](sparqlClient: SparqlClient[F], renkuUrl: RenkuUrl)
+      extends ProjectAuthService[F] {
+    private[this] val graph = Schemas.renku / "ProjectAuth"
+    private implicit val rUrl: RenkuUrl = renkuUrl
+
+    override def remove(slugs: NonEmptyList[Slug]): F[Unit] =
+      sparqlClient.update(sparql"""PREFIX schema: <http://schema.org/>
+                                  |PREFIX renku: <https://swissdatasciencecenter.github.io/renku-ontology#>
+                                  |
+                                  |DELETE { Graph ${graph.asSparql} {?s ?p ?o} }
+                                  |WHERE {
+                                  |  Graph ${graph.asSparql} {
+                                  |    ?s a schema:Project;
+                                  |       renku:slug ?slug;
+                                  |       ?p ?o.
+                                  |    VALUES(?slug) { ${slugs.toList.map(_.value)} }
+                                  |  }
+                                  |}
+          """.stripMargin)
+
+    override def update(data: ProjectAuthData): F[Unit] = {
+      val jsonld = NamedGraph.fromJsonLDsUnsafe(graph, data.asJsonLD)
+      remove(data.path) >> sparqlClient.upload(jsonld)
+    }
+
+    override def updateAll: Pipe[F, ProjectAuthData, Nothing] =
+      _.chunks
+        .evalTap(_.toNel.map(_.map(_.path)).map(remove).getOrElse(().pure[F]))
+        .map(chunk =>
+          chunk.toNel match { // TODO improve that ergonomics for NamedGraph in jsonld4s
+            case Some(nel) => NamedGraph.fromJsonLDsUnsafe(graph, nel.head.asJsonLD, nel.tail.map(_.asJsonLD): _*)
+            case None      => NamedGraph(graph, Seq.empty)
+          }
+        )
+        .evalMap(sparqlClient.upload)
+        .drain
+
+    override def getAll(chunkSize: Int): Stream[F, ProjectAuthData] =
+      streamAll(chunkSize)
+
+    private def streamAll(chunkSize: Int) =
+      Stream
+        .iterate(0)(_ + chunkSize)
+        .evalMap(offset => getChunk(chunkSize, offset))
+        .takeWhile(_.nonEmpty)
+        .flatMap(Stream.emits)
+        .groupAdjacentBy(_._1)
+        .map { case (slug, rest) =>
+          val members = rest.toList.flatMap(_._3)
+          val vis     = rest.head.map(_._2)
+          vis.map(v => ProjectAuthData(slug, members.toSet, v))
+        }
+        .unNone
+
+    private def getChunk(limit: Int, offset: Int) =
+      sparqlClient.queryDecode[(Slug, Visibility, Option[ProjectMember])](
+        sparql"""PREFIX schema: <http://schema.org/>
```

Comment:
Do you think it's better to query all the rows and group them on scala side? As you know we can also ask Jena to group them so then you transporting a single row per slug. Just a thought, though :)

Reply:
I tried both ways and couldn't see a performance gain. Then opted for this approach because there is no string manipulation. But not a strong opinion at all!

Reply:
Probably there won't be much of a difference. It's maybe just a matter which CPU, TG's or Jena's would do more cycles :)